391 documentation update (#392)
insight-platform/Savant: docs/source/advanced_topics/0_pipeline_benchmarking.rst

Benchmarking And Optimization
=============================

When a pipeline is ready, developers often need to measure its performance. There are three types of benchmarks developers want to run:

- non-real-time single stream performance;
- non-real-time multiple stream performance;
- real-time multiple stream performance.

There is one more dimension to benchmark:

- with adapters (end-to-end);
- without adapters (only pipeline).

> When benchmarking, we usually measure the aggregate throughput of the pipeline in FPS (frames-per-second) it can handle. Sometimes, usually, when the pipeline is intended to work in real-time, we also measure end-to-end latency, but this document does not discuss it.

Reviewer: ```suggestion
When benchmarking, we usually measure the aggregate throughput of the pipeline in FPS (frames-per-second). When the pipeline is intended to work in real-time and in some other cases, we may also measure end-to-end latency, but this document does not discuss it.
```

To measure the performance, use the ``uridecodebin`` Gstreamer source:

.. _isolated_uridecodebin_benchmark:

.. code-block::

    pipeline:
      # local video file source, not using source adapter
      source:
        element: uridecodebin
        properties:
          uri: file:///data/file.mp4

      # define pipeline's main elements
      elements:
        ...

      # noop pipeline sink, not using sink adapter
      sink:
        - element: devnull_sink

At the end of operation, you will see the FPS result.

Measuring Non-Real-Time Multiple Stream Performance
---------------------------------------------------

This kind of benchmarking is valuable to discover the maximum aggregate number of FPS the pipeline can reach. It may occur that the maximum value will be reached when ingesting 32 pipelines, but each of them will be processed at the rate of 5 FPS. That is why it is not real-time performance.

> To run such benchmarks we implemented a special adapter: :ref:`Multi-Stream Source Adapter <multi_stream_source_adapter>`, allowing the ingesting of a selected video file in parallel in the pipeline under benchmarking. By changing the number of parallel streams you may find out the value which is the maximum for the pipeline.

Reviewer: ```suggestion
To run such benchmarks we implemented a special adapter: :ref:`Multi-Stream Source Adapter <multi_stream_source_adapter>`, allowing the ingesting of a selected video file in parallel in the pipeline under benchmarking. By changing the number of parallel streams you can find out the maximum value for the pipeline.
```

To measure non-real-time performance with it, use ``SYNC_OUTPUT=False``.

Measuring Real-Time Multiple Stream Performance
-----------------------------------------------

This kind of benchmarking is valuable to find out the maximum aggregate number of FPS the pipeline can handle in real time. Considering the per-stream FPS is 30, you are looking for working configurations satisfying the equation ``N = X / 30``, where ``N`` is the number of streams ingested in the pipeline, and ``X`` is the aggregate FPS.

To run such benchmarks, you also can use the :ref:`Multi-Stream Source Adapter <multi_stream_source_adapter>` but set ``SYNC_OUTPUT=True``. By changing the number of parallel streams, you need to determine the value which is the maximum for the pipeline.

End-To-End or Isolated Benchmarking
-----------------------------------

In the above-mentioned :ref:`listing <isolated_uridecodebin_benchmark>` you may see that the sink for the pipeline is set to:

.. code-block::

    pipeline:
      # noop pipeline sink, not using sink adapter
      sink:
        - element: devnull_sink

> It represents benchmarking without real-life sinks which can form a bottleneck. If you want to test CV/ML performance it is what you are looking for. However, often you need to test end-to-end, including a specific sink implementation used practically. In such a situation, you need to include in the benchmark additional components, like a sink adapter you are planning to use and 3rd-party systems.

Reviewer: ```suggestion
It represents benchmarking without real-life sinks which can form a bottleneck. This is what you are looking for if you want to test CV/ML performance . However, often you need to test end-to-end, including a specific sink implementation used practically. In such a situation, you need to include additional components in the benchmark, like a sink adapter you are planning to use and 3rd-party systems.
```
